{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Case Studies in Bioinformatics\n",
    "_Cyril Matthey-Doret, December 2016_\n",
    "\n",
    "## Background\n",
    "\n",
    "Ontogeny is the development of an organism from egg fertilization to its mature state. It requires expression of different genes at each stage of development, associated with phenotypic changes. There is a lot of discussion on how these sets of genes differ between species. The earliest studies of embryogenesis performed by Karl Ernst Ritter von Baer in the 19th century suggested that the most similar state of development between related species, also known as the phylotypic stage, happens at the earliest stages of embryogenesis. This lead to what is called the funnel model. However, more recent work has shown that this convergence happens later, earliest stages being more different between species, giving rise to the hourglass model. It is now possible to test these models by combining gene expression data at different stages of embryogenesis and phylogenetic information for the concerned genes to better study the differences in expressed genes between species. Here, I investigate a study relying on an approach termed phylostratigraphy, which consists in tracing the origin of specific genes through similarity searches. This allows to build a phylogenetic tree where each gene has a rank that estimates their age. These ranks can be used to get a metric of age for the whole transcriptome at each developmental stage, called Transcriptome Age Index (TAI). Results obtained in this study support the hourglass model, focusing on expression datasets from the zebrafish _Danio rerio_, Drosophilia, the mosquito _Anopheles_ and the nematode _Caenorhabditis elegans_. In this report, I use forensic bioinformatics to repeat the analysis from [Domazet-Lošo et Tautz, 2010](http://www.nature.com/nature/journal/v468/n7325/full/nature09632.html) to have a critical view of each step and a better understanding of their results. Finally, I propose improvements for the analysis, explaining how it would affect the results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods\n",
    "\n",
    "I reproduced the analysis from the paper using python 2 with the modules panda, numpy and GEOparse. I used expression data from _Danio rerio_ (zebrafish) to reproduce the results shown in figure 1a of the paper. Expression data for zebrafish was obtained from [Domazet-Lošo et Tautz (2010)](http://www.nature.com/nature/journal/v468/n7325/full/nature09632.html) and phylostrata were obtained from [Šestak et al, 2013](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3636138/). The first step was to format the expression data from microarray and find the gene ID corresponding to the spots ID in the microarray. This information can be obtained via the GEO platform (GPL) annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "hideOutput": false,
    "hidePrompt": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing ./GSE24616.soft.gz:\n",
      " - DATABASE : GeoMiame\n",
      " - SERIES : GSE24616\n",
      " - PLATFORM : GPL6457\n",
      " - SAMPLE : GSM606866\n",
      " - SAMPLE : GSM606867\n",
      " - SAMPLE : GSM606868\n",
      " - SAMPLE : GSM606869\n",
      " - SAMPLE : GSM606870\n",
      " - SAMPLE : GSM606871\n",
      " - SAMPLE : GSM606872\n",
      " - SAMPLE : GSM606873\n",
      " - SAMPLE : GSM606874\n",
      " - SAMPLE : GSM606875\n",
      " - SAMPLE : GSM606876\n",
      " - SAMPLE : GSM606877\n",
      " - SAMPLE : GSM606878\n",
      " - SAMPLE : GSM606879\n",
      " - SAMPLE : GSM606880\n",
      " - SAMPLE : GSM606881\n",
      " - SAMPLE : GSM606882\n",
      " - SAMPLE : GSM606883\n",
      " - SAMPLE : GSM606884\n",
      " - SAMPLE : GSM606885\n",
      " - SAMPLE : GSM606886\n",
      " - SAMPLE : GSM606887\n",
      " - SAMPLE : GSM606888\n",
      " - SAMPLE : GSM606889\n",
      " - SAMPLE : GSM606890\n",
      " - SAMPLE : GSM606891\n",
      " - SAMPLE : GSM606892\n",
      " - SAMPLE : GSM606893\n",
      " - SAMPLE : GSM606894\n",
      " - SAMPLE : GSM606895\n",
      " - SAMPLE : GSM606896\n",
      " - SAMPLE : GSM606897\n",
      " - SAMPLE : GSM606898\n",
      " - SAMPLE : GSM606899\n",
      " - SAMPLE : GSM606900\n",
      " - SAMPLE : GSM606901\n",
      " - SAMPLE : GSM606902\n",
      " - SAMPLE : GSM606903\n",
      " - SAMPLE : GSM606904\n",
      " - SAMPLE : GSM606905\n",
      " - SAMPLE : GSM606906\n",
      " - SAMPLE : GSM606907\n",
      " - SAMPLE : GSM606908\n",
      " - SAMPLE : GSM606909\n",
      " - SAMPLE : GSM606910\n",
      " - SAMPLE : GSM606911\n",
      " - SAMPLE : GSM606912\n",
      " - SAMPLE : GSM606913\n",
      " - SAMPLE : GSM606914\n",
      " - SAMPLE : GSM606915\n",
      " - SAMPLE : GSM606916\n",
      " - SAMPLE : GSM606917\n",
      " - SAMPLE : GSM606918\n",
      " - SAMPLE : GSM606919\n",
      " - SAMPLE : GSM606920\n",
      " - SAMPLE : GSM606921\n",
      " - SAMPLE : GSM606922\n",
      " - SAMPLE : GSM606923\n",
      " - SAMPLE : GSM606924\n",
      " - SAMPLE : GSM606925\n",
      " - SAMPLE : GSM606926\n",
      " - SAMPLE : GSM606927\n",
      " - SAMPLE : GSM606928\n",
      " - SAMPLE : GSM606929\n",
      " - SAMPLE : GSM606930\n",
      " - SAMPLE : GSM606931\n",
      " - SAMPLE : GSM606932\n",
      " - SAMPLE : GSM606933\n",
      " - SAMPLE : GSM606934\n",
      " - SAMPLE : GSM606935\n",
      " - SAMPLE : GSM606936\n",
      " - SAMPLE : GSM606937\n",
      " - SAMPLE : GSM606938\n",
      " - SAMPLE : GSM606939\n",
      " - SAMPLE : GSM606940\n",
      " - SAMPLE : GSM606941\n",
      " - SAMPLE : GSM606942\n",
      " - SAMPLE : GSM606943\n",
      " - SAMPLE : GSM606944\n",
      " - SAMPLE : GSM606945\n",
      " - SAMPLE : GSM606946\n",
      " - SAMPLE : GSM606947\n",
      " - SAMPLE : GSM606948\n",
      " - SAMPLE : GSM606949\n",
      " - SAMPLE : GSM606950\n",
      " - SAMPLE : GSM606951\n",
      " - SAMPLE : GSM606952\n",
      " - SAMPLE : GSM606953\n",
      " - SAMPLE : GSM606954\n",
      " - SAMPLE : GSM606955\n",
      " - SAMPLE : GSM606956\n",
      " - SAMPLE : GSM606957\n",
      " - SAMPLE : GSM606958\n",
      " - SAMPLE : GSM606959\n",
      " - SAMPLE : GSM606960\n",
      " - SAMPLE : GSM606961\n",
      " - SAMPLE : GSM606962\n",
      " - SAMPLE : GSM606963\n",
      " - SAMPLE : GSM606964\n",
      " - SAMPLE : GSM606965\n",
      " - SAMPLE : GSM606966\n",
      " - SAMPLE : GSM606967\n",
      " - SAMPLE : GSM606968\n",
      " - SAMPLE : GSM606969\n",
      " - SAMPLE : GSM606970\n",
      " - SAMPLE : GSM606971\n",
      " - SAMPLE : GSM606972\n",
      " - SAMPLE : GSM606973\n",
      " - SAMPLE : GSM606974\n",
      " - SAMPLE : GSM606975\n",
      " - SAMPLE : GSM606976\n",
      " - SAMPLE : GSM606977\n",
      " - SAMPLE : GSM606978\n",
      " - SAMPLE : GSM606979\n",
      " - SAMPLE : GSM606980\n",
      " - SAMPLE : GSM606981\n",
      " - SAMPLE : GSM606982\n",
      " - SAMPLE : GSM606983\n",
      " - SAMPLE : GSM606984\n",
      " - SAMPLE : GSM606985\n",
      " - SAMPLE : GSM606986\n",
      " - SAMPLE : GSM606987\n",
      " - SAMPLE : GSM606988\n",
      " - SAMPLE : GSM606989\n",
      " - SAMPLE : GSM606990\n",
      " - SAMPLE : GSM606991\n",
      " - SAMPLE : GSM606992\n",
      " - SAMPLE : GSM606993\n",
      " - SAMPLE : GSM606994\n",
      " - SAMPLE : GSM606995\n",
      " - SAMPLE : GSM606996\n",
      " - SAMPLE : GSM606997\n",
      " - SAMPLE : GSM606998\n",
      " - SAMPLE : GSM606999\n",
      " - SAMPLE : GSM607000\n",
      " - SAMPLE : GSM607001\n",
      " - SAMPLE : GSM607002\n",
      " - SAMPLE : GSM607003\n",
      " - SAMPLE : GSM607004\n",
      " - SAMPLE : GSM607005\n",
      " - SAMPLE : GSM607006\n",
      " - SAMPLE : GSM607007\n",
      " - SAMPLE : GSM607008\n",
      " - SAMPLE : GSM607009\n",
      " - SAMPLE : GSM607010\n",
      " - SAMPLE : GSM607011\n",
      " - SAMPLE : GSM607012\n"
     ]
    }
   ],
   "source": [
    "#!pip install GEOparse\n",
    "\n",
    "import matplotlib.pylab as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import GEOparse\n",
    "\n",
    "probes_conv = GEOparse.parse_GSM(\"/home/cyril/Documents/Master/sem_1/Case_study/module1/data/GPL6457_old_annotations.txt.gz\")\n",
    "\n",
    "try:\n",
    "    gse = GEOparse.get_GEO(filepath=\"./GSE24616.soft.gz\", silent=True)\n",
    "except IOError:\n",
    "    gse = GEOparse.get_GEO(\"GSE24616\",destdir=\"./\", silent=True)\n",
    "# Using GEOparse inbuilt tools to parse the Gene series expression data for zebrafish.\n",
    "# If the file is not found, it is downloaded first, otherwise it is parsed directly\n",
    "\n",
    "char = {\"stage\": [], \"time\": [], \"sex\": [], \"sample_name\": []}\n",
    "# Initializing dictionary object to store metadata\n",
    "\n",
    "for gsm_name, gsm in sorted(gse.gsms.iteritems()): \n",
    "    char[\"stage\"].append(gsm.metadata['characteristics_ch1'][1].split(\": \")[1])\n",
    "    char[\"time\"].append(gsm.metadata['characteristics_ch1'][2].split(\": \")[1])\n",
    "    char[\"sex\"].append(gsm.metadata['characteristics_ch1'][3].split(\": \")[1])\n",
    "    char[\"sample_name\"].append(gsm.name)\n",
    "# Formatting parsed metadata in a structured dictionary object\n",
    "\n",
    "GPL = gse.gpls.values()[0]\n",
    "pivoted_samples = gse.pivot_samples('VALUE')\n",
    "pivoted_samples.set_index(GPL.table.SPOT_ID, inplace=True)\n",
    "# Setting indexes as microarray spot ID\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, I imported the phylostrata and merged it with expression data using probes ID. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "\n",
    "strata = pd.read_csv(\"./phylostrata.txt\", sep=\"\\t\", header=None)\n",
    "strata.columns = [\"GeneID\", \"ProbeID\", \"age\"]\n",
    "strata.set_index(\"ProbeID\", inplace=True)\n",
    "# Loading and formatting phylostrata with matching genes ID, probe ID and gene age\n",
    "\n",
    "matched_data = pivoted_samples.join(strata, how=\"inner\").groupby(level=0).last()\n",
    "# Joining expression data with phylostrata by probe ID\n",
    "\n",
    "unique_data = matched_data.groupby(\"GeneID\").mean()\n",
    "# Taking the mean expression values for all genes.\n",
    "\n",
    "char_pd = pd.DataFrame(char, index=char[\"sample_name\"])\n",
    "mixed = char_pd[char_pd.sex == \"mixed\"].sample_name.tolist()\n",
    "mixed += char_pd[char_pd.sex == \"female\"].sample_name.tolist()\n",
    "\n",
    "char_pd[\"timing_number\"] = 0\n",
    "time_stamps = char_pd.time.unique()\n",
    "for i in xrange(len(time_stamps)):\n",
    "    char_pd.loc[char_pd.time == time_stamps[i], \"timing_number\"] = i + 1\n",
    "\n",
    "    experiment_index = char_pd[char_pd.index.isin(mixed)].reset_index().groupby(\"timing_number\")[\"index\"].apply(\n",
    "        lambda x: np.array(x))\n",
    "\n",
    "set_mean = {}\n",
    "stages = []\n",
    "for d, col_list in experiment_index.iteritems():\n",
    "    set_mean[d] = unique_data[col_list].mean(axis=1)\n",
    "    stages.append(char_pd[char_pd.index.isin(col_list)].stage[0])\n",
    "\n",
    "mean_data = pd.DataFrame(set_mean)\n",
    "mean_data.plot()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "References: \n",
    "\n",
    "Šestak, M. S., Božičević, V., Bakarić, R., Dunjko, V., & Domazet-Lošo, T. (2013). Phylostratigraphic profiles reveal a deep evolutionary history of the vertebrate head sensory systems. Frontiers in Zoology, 10, 18. http://doi.org/10.1186/1742-9994-10-18\n",
    "\n",
    "Domazet-Lošo T, Tautz D. (2010). A phylogenetically based transcriptome age index mirrors ontogenetic divergence patterns. Nature, 468(7325):815-8. http://doi.org/10.1038/nature09632\n",
    "\n",
    "Data files: \n",
    "\n",
    "Zebrafish expression data: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE24616\n",
    "\n",
    "Phylostrata: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3636138/bin/1742-9994-10-18-S5.xls\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
